### ISSUE TITLE
Question about  function box_candidates() in datasets.py

### ISSUE THREAD

**Issue by TTLuze** [ultralytics/yolov5#2442]
Hello,

I think this is a mistake. At the end of function random_perspective(), box_candidates() is used to filters the bbox based on the the area_ratio of box2 to box1.
`
        i = box_candidates(box1=targets[:, 1:5].T * s, box2=new.T, area_thr=0.01 if use_segments else 0.10)
`

While in load_mosaic(), the boxes are cliped before being sent to random_perspective(), which might cut the bbox out of the mosaic images. It makes the area of box1 smaller. And some box2 would not be filtered even only a small part of it remains in the image.
```    
for x in (labels4[:, 1:], *segments4):
    np.clip(x, 0, 2 * s, out=x)  # clip when using random_perspective()
```

By the way, for the filtered box, is masking the area with (114, 114, 114) a better method rather than ignore it?

**Comment by glenn-jocher** [comment]
@TTLuze yes box_candidates() is used to filter the labels and reject poor label candidates:
https://github.com/ultralytics/yolov5/blob/886f1c03d839575afecb059accf74296fad395b6/utils/datasets.py#L932-L938

`wh_thr`=2 means that boxes must have widths and heights larger than 2 pixels, and ar_thr=20 means that boxes must have aspect ratios (width/height or height/width) less than 20 and greater than 1/20.

`area_thr`=0.10 means that boxes must have 10% of their original area left after augmentation, adjusted for scale differences. So if a box is 95% cropped by augmentation, then it will not be a box candidate.

load_mosaic() clipping does not remove boxes, it may clip their values, reducing their area or even leaving them with zero area. 

Zero area boxes and very small boxes are rejected by the `wh_thr` criteria.

**Comment by glenn-jocher** [comment]
@TTLuze perhaps you are recommending an additional box_candidate call in load_mosaic() after clipping to reject boxes that are nearly entirely clipped?

**Comment by TTLuze** [comment]
> @TTLuze perhaps you are recommending an additional box_candidate call in load_mosaic() after clipping to reject boxes that are nearly entirely clipped?

yes. That is my point. It is better to filter the boxes using original unclipped size , or call box_candidate() after every clipping operation.
 
By the way, there is another question. Yolov5 simply remove the poor label candidates, but leave the image unchanged. Would it be better to mask the filtered boxe swith (114, 114, 114) ?

**Comment by glenn-jocher** [comment]
@TTLuze then you might try experimenting with applying box_candidates() first in mosaic and then in random_perspective. The reason the labels are clipped is that otherwise the augmentation operations in random_perspective might expose those label areas that fell outside of the mosaic, so that the label would extend out over grey area beyond the original mosaic. 

TLDR we can not remove the clipping in mosaic, it is necessary, but we can apply box_candidates also in mosaic function.

I'm not sure what effect greying rejected boxes would have, probably little impact either way. Only way to know is to run that experiment.

**Comment by adamhrv** [comment]
@glenn-jocher would you accept PR for adding `area_thr` to YAML configs? Can also add other args for `box_candidates`. Acceptable box candidate thresholds can vary for objects in custom datasets, especially on asymmetric objects.

**Comment by glenn-jocher** [comment]
@adamhrv we are always happy to review PRs! 

For all of the hyps already included, there are many more that are still hardcoded as you mentioned, such as box candidates and NMS settings.
